external-crates/move/crates/move-compiler/src/sui_mode/linters/missing_key.rs

```diff
@@ -0,0 +1,65 @@
+//! This linter rule checks for structs with an `id` field of type `UID` without the `key` ability.
+//! It generates a warning if a struct incorrectly models unique identifiers without required abilities.
+use crate::{
+    diag,
+    diagnostics::codes::{custom, DiagnosticInfo, Severity},
+    linters::MISSING_KEY_DIAG_CODE,
+    naming::ast::{StructDefinition, StructFields, TypeName_, Type_},
+    parser::ast::Ability_,
+    shared::CompilationEnv,
+    typing::{ast as T, visitor::TypingVisitor},
+};
+use move_ir_types::location::Loc;
+
+use super::{LinterDiagnosticCategory, LINT_WARNING_PREFIX};
+
+const MISSING_KEY_ABILITY_DIAG: DiagnosticInfo = custom(
+    LINT_WARNING_PREFIX,
+    Severity::Warning,
+    LinterDiagnosticCategory::Sui as u8,
+    MISSING_KEY_DIAG_CODE,
+    "Struct has an 'id' field of type 'UID' but is missing the 'key' ability.",
+);
+
+pub struct MissingKey;
+
+impl TypingVisitor for MissingKey {
```

Comment:
This needs to use TypingVisitorConstructor and TypingVisitorContext or it is going to skip the lint/warning suppression

Reply:
refactored